app/models/reports/drug_stock_calculation.rb

```diff
@@ -0,0 +1,159 @@
+module Reports
+  class DrugStockCalculation
+    def initialize(state:,
+      protocol:,
+      drug_category:,
+      stocks_by_rxnorm_code:,
+      patient_count: nil,
+      previous_month_stocks_by_rxnorm_code: nil)
+      @protocol = protocol
+      @drug_category = drug_category
+      @stocks_by_rxnorm_code = stocks_by_rxnorm_code
+      @previous_month_stocks_by_rxnorm_code = previous_month_stocks_by_rxnorm_code
+      @patient_count = patient_count
+      @coefficients = patient_days_coefficients(state)
+    end
+
+    def protocol_drugs_by_category
+      @protocol_drugs_by_category ||= @protocol.protocol_drugs.where(stock_tracked: true).group_by(&:drug_category)
+    end
+
+    def patient_days
+      return nil if stocks_on_hand.empty?
+      {stocks_on_hand: stocks_on_hand,
+       patient_count: @patient_count,
+       load_coefficient: @coefficients[:load_coefficient],
+       new_patient_coefficient: new_patient_coefficient,
+       estimated_patients: estimated_patients,
+       patient_days: patient_days_calculation}
+    rescue => e
```

Comment:
@kpethtel You're right, we should try to catch specific error types here if we can. Agree with the general principles outlined in your comment.

I'll add some context around why we started off with a catch-all so we understand the rationale for the way things are right now. There were a number of potential issues we would run into with the calculations here:
- The formulae for stock calculations are represented in YAML, which is manually maintained. It's possible that we misconfigured something there for a particular state. This might raise issues in accessing parts of a hash, say.
- The protocol drugs being tracked might not be the same drugs that are present in the formula. Working this into the code isn't worth it at this point in time since it adds significant complexity. And we're looking to move towards first-class formulae at the protocol level anyway so this would go away quickly.
- The stock, or coefficients might not be numeric (string, nil, etc), so we run into arithmetic exceptions
- We might divide by 0

The requirement is that we show the drug stock report on the dashboard despite errors on individual rows, or per drug category. And the calculations that have errors need to show up as "error" in the table cell, so that we have feedback and can then dig into why that's the case. The feedback might also be to the CVHO, etc to fix their data accordingly. That's the reason why we're catching the error here, rather than one level lower say in `stocks_on_hand `, since we need to report it as an error overall if this calculation fails.

However, we are already handling some cases of `nil`s where we are aware of the scenario. And at this point, we might be confident enough in the domain to make choices about what errors we catch. I propose that we take a shot at this, separate from this PR.

Reply:
Thanks for explaining @ssrihari. It sounds like this is a complex project with a lot of unknowns. My biggest concern is that it's not clear to me exactly what errors we're guarding against. Clarifying that will make this easier code to understand. I think making error handling improvements in a future ticket is reasonable. Can we make a ticket for that or add it as a requirement to an existing ticket?

Reply:
Added a follow on ticket for this, as well as other improvements. Will attempt to wrap up this week: https://app.clubhouse.io/simpledotorg/story/2806/drug-consumption-improvements